### PR TITLE
ar71xx: add support new revision of Mikrotik hardware

### PR DESCRIPTION
**Pull request by NewUse**
New algorithm of finding radio calibration data to  support new revisions of Mikrotik hardware, 
tested by Lipton2kirov with both rb951ui-2hnd rev1 and rev2, ported from: 
https://github.com/robimarko/openwrt/blob/BaseBox2/package/utils/rbextract/src/rbextract.c
support discussion:
https://forum.openwrt.org/t/add-support-for-mikrotik-rb912uag-2hpnd-r2/45215
and
https://forum.openwrt.org/t/support-for-mikrotik-rb951ui-2hnd-r2/46923

PS: I am not a good programmer, so patch need to be reviewed.
PPS: is not working in 19.x and master branches

### CODE REVIEW

**Comment by NewUse** [comment]
New algorithm of finding radio colibration data to  support new revisions of Mikrotik hardware, tested by Lipton2kirov with both rb951ui-2hnd rev1 and rev2, ported from: 
https://github.com/robimarko/openwrt/blob/BaseBox2/package/utils/rbextract/src/rbextract.c
support discussion:
https://forum.openwrt.org/t/add-support-for-mikrotik-rb912uag-2hpnd-r2/45215
and
https://forum.openwrt.org/t/support-for-mikrotik-rb951ui-2hnd-r2/46923

PS: I am not a good programmer, so patch need to be rewieved.
PPS:  Not working in 19.* or master branches

**Comment by adrianschmutzler** [comment]
Hi, your PR has formal issue, please read https://openwrt.org/submitting-patches again.

> PPS: is not working in 19.x and master branches

Then please fix it there. Your patch will first have to go into master, and then be backported to 19.07 and 18.06.
However, be aware that ar71xx is a deprecated target that will not be supported beyond 19.07. Some people (e.g. @rogerpueyo ) are currently working on migrating Mikrotik to ath79. Given the size of your contribution, it looks to me like it would be more useful to build that directly into new ath79, as implementing it into ar71xx will consume a lot of time and it will be hard to find somebody wanting to review this as well.

After all, you might get this into ar71xx only a few months before a 20.xx release and the EOL of 18.06 ...

@xback FYI

**Comment by adrianschmutzler** [comment]
Ah, and since this is based on 18.06, I will close it for now due to the reasons described above. Please provide a patch that can be applied to master.
Please don't feel repelled by that, it's just the way how development works here :-)

**Comment by rogerpueyo** [comment]
Hi @NewUse,

As Adrian has stated, it is much better to focus efforts in making things work for master (or 19.07, at least) and, better than adding things to ar71xx, it would be more desirable to add them to ath79.

Regarding the code you added, it is already in master for the ath79 generic (https://github.com/openwrt/openwrt/commit/ba730d61af3b860e5d5a086a318a6c2a84e68c9c), plus some later modifications to support older devices (https://github.com/openwrt/openwrt/commit/7d39946ea0fe954bcd0c4d52fffc6d9dc6248cf6) which might be useful for adding support to the rb951ui-2hnd r2 in ath79. For this, you may want to use this https://github.com/rogerpueyo/openwrt/tree/ath79-mikrotik-rb-922uags-5hpacd as an example.

**Comment by NewUse** [comment]
Thanks, it seems that this pull-request  is real useless, I have fixed in my fork one mistype, and now master is also working, but it would be enough to port RB to new ath79 instead of this fix, I'll try to add this support in a few days but I am not sure : )

**Comment by xback** [comment]
This still makes sense for 19.07
I tried it today on a rb912 r2 board and it definitely fixed the wifi-missing issue.

If it's ok for you, I'll check if this is the best way for 19.07 (after cleaning it up)

Thanks!

**Comment by xback** [comment]
something is still wrong with the patch.
The device is not properly detected:

```
[    0.138902] MIPS: machine is Mikrotik RouterBOARD 
[    0.375043] rb: Decompressing with LZO
```

**Comment by NewUse** [comment]
Sorry I am new to git and github, I don't know how correctly add changes to exists Pull Request

**Comment by NewUse** [comment]
Could you, please provide more info from logs and H/W revisions of routers you are using, this patch was successfully tested by Lipton2kirov with both 1st and 2nd revisions.

**Comment by xback** [comment]
@NewUse 
Sure, I'm testing it on 19.07 using RB912UAG-5HPnD r2

The problem is that the initial flash works as expected, but afterwards the board is not properly detected anymore.
This leads to issues like:
- not able to use sysupgrade afterwards
- the PCIe slot doesn't work anymore
- USB doesn't work anymore
- ...

```
[ Node 3 | node-3 ] cat /etc/board.json 
{
	"model": {
		"id": "unknown",
		"name": "Mikrotik RouterBOARD "
	},
	"network": {
		"lan": {
			"ifname": "eth0",
			"protocol": "static"
		},
		"wan": {
			"ifname": "eth1",
			"protocol": "dhcp"
		}
	}
}

```

**Comment by NewUse** [comment]
Seems to be 912-specific, just tested by Lipton2kirov with 951 (new revision) and openwrt 19.07:
root@OpenWrt:~# cat /etc/board.json 
{
  "model": {
    "id": "rb-951ui-2hnd",
    "name": "MikroTik RouterBOARD 951Ui-2HnD"
  },
  "network": {
    "lan": {
      "ifname": "eth1",
      "protocol": "static"
    },
    "wan": {
      "ifname": "eth0",
      "protocol": "dhcp"
    }
  }
}
root@OpenWrt:~# cat /tmp/sysinfo/model 
MikroTik RouterBOARD 951Ui-2HnD
root@OpenWrt:~# cat /tmp/sysinfo/board_name 
rb-951ui-2hnd
May be specific entry to /target/linux/ar71xx/base-files/lib/ar71xx.sh should be added, unfortunately I don't have actual dump of calibration data of rb912 new revision, and can't see what RB_ID_BOARD_NAME is there set.

**Comment by Lipton2kirov** [comment]
@xback you can show cat /tmp/sysinfo/model && cat /tmp/sysinfo/board_name && cat /proc/cpuinfo?

**Comment by xback** [comment]
@Lipton2kirov 
Sure:

[ Node 3 | node-3 ] cat /tmp/sysinfo/model
Mikrotik RouterBOARD 
[ Node 3 | node-3 ] cat /tmp/sysinfo/board_name
unknown
[ Node 3 | node-3 ] cat /proc/cpuinfo
system type		: Atheros AR9342 rev 3
machine			: Mikrotik RouterBOARD 
processor		: 0
cpu model		: MIPS 74Kc V4.12
BogoMIPS		: 299.82
wait instruction	: yes
microsecond timers	: yes
tlb_entries		: 32
extra interrupt vector	: yes
hardware watchpoint	: yes, count: 4, address/irw mask: [0x0ffc, 0x0ffc, 0x0ffb, 0x0ffb]
isa			: mips1 mips2 mips32r1 mips32r2
ASEs implemented	: mips16 dsp dsp2
Options implemented	: tlb 4kex 4k_cache prefetch mcheck ejtag llsc dc_aliases perf_cntr_intr_bit nan_legacy nan_2008 perf
shadow register sets	: 1
kscratch registers	: 0
package			: 0
core			: 0
VCED exceptions		: not available
VCEI exceptions		: not available

[ Node 3 | node-3 ]

**Comment by xback** [comment]
The issue seems to be that it starts correctly, but finished to early reading the data ..

Expected:

machine			: Mikrotik RouterBOARD 912UAG-5HPnD

**Comment by Lipton2kirov** [comment]
@xback When building, are these files used?
https://github.com/openwrt/openwrt/blob/master/target/linux/ar71xx/files/arch/mips/ath79/mach-rb91x.c
https://github.com/openwrt/openwrt/blob/master/target/linux/ar71xx/base-files/lib/ar71xx.sh

**Comment by xback** [comment]
@Lipton2kirov 

Here is the reproducer:

- Clone and checkout 19.07
- Build ar71xx with nand for Mikrotik without AC
- Flash and run
--> All is OK

- Clone and checkout 19.07
- Apply this series
- Build ar71xx with nand for Mikrotik without AC
- Flash and run
--> Board detect fails

**Comment by NewUse** [comment]
I am not sure, it seems to be 911 mach-file specific, I have made very small changes but I am absolutely not sure that I did correctly understand algorithm of mach search and board naming.

**Comment by NewUse** [comment]
seems that it's compiles without errors, @xback please, test it.

File: target/linux/ar71xx/files/arch/mips/ath79/mach-rb91x.c [hunk 1]
```diff
@@ -340,4 +335,4 @@ static void __init rb711gr100_setup(void)
 
 }
 
-MIPS_MACHINE_NONAME(ATH79_MACH_RB_711GR100, "711Gr100", rb711gr100_setup);
+MIPS_MACHINE(ATH79_MACH_RB_711GR100, "711Gr100", "MikroTik RouterBOARD 912UAG-5HPnD", rb711gr100_setup);
```
Comment (f00b4r0) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-rb91x.c]: This is wrong: it break all other RB devices that relied on that code (e.g. RB911G-*) which will now be mis-identified.
See https://github.com/openwrt/openwrt/blob/494babf34c7a41cf1a1775530ba0d2e4ebc0c43f/target/linux/ar71xx/files/arch/mips/ath79/mach-rb91x.c#L264

Comment (NewUse) [reply target/linux/ar71xx/files/arch/mips/ath79/mach-rb91x.c]: Yes, sure, you are right, but first of all I wanna know if the trouble is in wrong mach detection of new revision, or somewhere else. I am not an expert of openwrt code and unfortunately I don't have the 91x new revision hardware to test myself. If this fix will help, then it could be possible to rewrite it to detect all other models.

Comment (f00b4r0) [reply target/linux/ar71xx/files/arch/mips/ath79/mach-rb91x.c]: The problem is this is not a fix. What it does is statically declare that all machines booting with "711Gr100" board model will be detected and registered as "MikroTik RouterBOARD 912UAG-5HPnD". There's no way to "rewrite it to detect all other models". The previous code did exactly that.

The original code pulled the actual device name from the flash. As far as I know, all mikrotik devices store their model name in flash (in hard_config), it would be interesting to find out why this doesn't work with yours. The dumps provided by @xback above do contain that data.

HTH

Comment (NewUse) [reply target/linux/ar71xx/files/arch/mips/ath79/mach-rb91x.c]: Thanks alot, I have compared 951 and 912 hard_configs and I think I have found my error in routerboard.c file in detection in rb_get_board_name function, I'll correct it at Monday. Thanks! You where right, there is no need to change mach-rb91x file.

**Comment by NewUse** [comment]
@xback , @Lipton2kirov, please, test again with both old and new revisions, if ig is possible.

File: target/linux/generic/files/include/linux/routerboot.h [hunk 2]
```diff
@@ -42,6 +45,9 @@
 #define RB_ID_MAC_ADDRESS_COUNT	14
 #define RB_ID_HW_OPTIONS	21
 #define RB_ID_WLAN_DATA		22
+#define RB_ID_BOARD_IDENTIFIER	23
+#define RB_ID_BOARD_NAME	33
```
Comment (f00b4r0) [inline target/linux/generic/files/include/linux/routerboot.h]: I'm curious why RB_ID_BOARD_NAME was moved from 5 to 33. Maybe @robimarko can comment?
In order to preserve old behavior, rb_get_board_name() should try the old ID first I think.

Comment (NewUse) [reply target/linux/generic/files/include/linux/routerboot.h]: Yes, I think, we should ask, but there was no 33 parameter in old hard_config revisions, but it seems that there could present both 05 and 33 in new revision, that's why we should try 33 first and return it if it is present. So, I think that my fix is savefull and both backward and forward compatible...

**Comment by xback** [comment]
for completeness:

I backported rbextract to 19.07 locally here:

```
[ node-3 ] rbextract -a /dev/mtd1
Board name: 
Board product code: 912UAG-5HPnD
Board identifier: 711Gr100
Board serial: A2860BFDA4ED
Board MAC: c4:ad:34:08:89:48
HW Options 200205
Factory RouterBoot version: 6.44.5
Flash identifier: c81040c8
```

File: target/linux/generic/files/include/linux/routerboot.h (hunk 2 above)
Comment (xback) [reply target/linux/generic/files/include/linux/routerboot.h]: fyi,

R1:
```
[ Nobelwind OHVS 1 | node-13 ] strings /dev/mtd1
draH
l;kB
77BA06BD22F5
912UAG-5HPnD
3.41
711Gr100
tDDD
42064206420
ABEG1257pu
<|<|<|<|<|<|<|<|<|<|<|<|<|<|<|<|<|<|<|<|<|<|<|<| 
LTl|
@FH068LThx
LThx
LThx
LThx
LThx
LThx
LThx
LThx
LThx
<|<|<|<|<|<|<|<|<|<|<|<|<|<|<|<|<|<|<|<|<|<|<|<
|<|<|<|<|<|<|<|<|<|<|<|<|
```


R2:
```
[ node-3 ] strings /dev/mtd1
draH
A2860BFDA4ED
912UAG-5HPnD
6.44.5
711Gr100
tDDD
4206
6662*662,
ABEG1257pu
?K*4
0LTl|
	N^f
6620<
U=.l
6l? 
642.l
@FH068LThx
```

**Comment by xback** [comment]
> @xback , @Lipton2kirov, please, test again with both old and new revisions, if it is possible.
> 
> Many thanks to f00b4r0 for pointing to possible troubles.

Not there yet:

```
[    0.138878] MIPS: machine is Mikrotik RouterBOARD 
[    0.375039] rb: Decompressing with LZO
```

**Comment by NewUse** [comment]
please, provide 
cat /tmp/sysinfo/model
cat /tmp/sysinfo/board_name
cat /proc/cpuinfo

and full boot log again : )  
Are there troubles with both R1 and R2? or only 2nd revision is not working?

File: target/linux/ar71xx/files/arch/mips/ath79/routerboot.c [hunk 3]
```diff
@@ -329,6 +329,8 @@ rb_get_board_name(void)
 	int err;
 
 	err = rb_find_hard_cfg_tag(RB_ID_BOARD_NAME, &tag, &tag_len);
+	if (err)
+		err = rb_find_hard_cfg_tag(RB_ID_BOARD_PRODUCT_CODE, &tag, &tag_len);
 	if (err)
 		return NULL;
```
Comment (xback) [inline target/linux/ar71xx/files/arch/mips/ath79/routerboot.c]: I added some prints showing the data after each call to `rb_find_hard_cfg_tag`

It seem that RB_ID_BOARD_NAME is present but does not contain valid data.
Hence the Err is not triggered and RB_ID_BOARD_PRODUCT_CODE is never checked.

```
[    0.138864] rb: board name: tag_len: 4  tag: 
[    0.143811] MIPS: machine is Mikrotik RouterBOARD 
[    0.389615] rb: Decompressing with LZO
```

**Comment by xback** [comment]
@NewUse 

I copied rb_find_hard_cfg_tag into a userspace app and checked it with a debugger both on a rb912 and rb922 which came from a netmetal 5.

It seems that each tag always has 4 bytes of data. even when it's empty.

RB922:
Size `00 0c` Type `00 21` payload `4e 65 74 4d  65 74 61 6c 20 35 00 00`  |...!NetMetal 5..|

RB912:
Size `00 04` Type `00 21` Payload `00 00 00 00`


Also, on a RB922 which comes from within a netmetal 5 housing:
[    0.126391] MIPS: machine is MikroTik RouterBOARD 922UAGS-5HPacD

So original ar71xx uses tag 05,  and not tag 33 to determine the board.
So the proper fix is to use tag 05 for board ID.

**Comment by f00b4r0** [comment]
That bug was basically introduced by moving RB_ID_BOARD_NAME from 05 to 33.

While I understand the reason (05 really is the board identifier - always available - 33 is the marketing name - not always available), since rb_get_board_name() still picks from RB_ID_BOARD_NAME, it gets the wrong result; even when 33 is available: the script expect the board identifier, not the marketing name.

If the desired behavior is to prioritize name over identifier, a lot of testing and work will have to be done to make sure no other board break (unlikely).

I'm working on a clean driver to present all that data and wlan calibration data in sysfs in a consistent way (so that routerboot.c and rbextract become unnecessary and userland can expect a stable interface): I will submit a patch for review soon.

HTH

**Comment by xback** [comment]
Confirmed that using Tag 05 fixes all issues

```
[    0.138867] MIPS: machine is Mikrotik RouterBOARD 912UAG-5HPnD
[    0.385871] rb: Decompressing with LZO

```

**Comment by xback** [comment]
> That bug was basically introduced by moving RB_ID_BOARD_NAME from 05 to 33.
> 
> While I understand the reason (05 really is the board identifier - always available - 33 is the marketing name - not always available), since rb_get_board_name() still picks from RB_ID_BOARD_NAME, it gets the wrong result; even when 33 is available: the script expect the board identifier, not the marketing name.
> 
> If the desired behavior is to prioritize name over identifier, a lot of testing and work will have to be done to make sure no other board break (unlikely).
> 
> I'm working on a clean driver to present all that data and wlan calibration data in sysfs in a consistent way (so that routerboot.c and rbextract become unnecessary and userland can expect a stable interface): I will submit a patch for review soon.
> 
> HTH

Cool!, cant wait for your patch :-)

`05` should be `Board_Name`   (actual physical board)
`33` should be `Product_Name` (Product the board is being used in)

**Comment by NewUse** [comment]
If I do understand completely, the real board_name is stored in RB_ID_BOARD_IDENTIFIER but unfortunately for many router models could have the same main board, I have not looked deeply in code, but maybe the correct way is to use RB_ID_BOARD_IDENTIFIER and use the same params(with all available hardware) for all boards with the same RB_ID_BOARD_IDENTIFIER and the RB_ID_BOARD_PRODUCT_CODE only for displaying the marketing information. but this could affect  too many mach-files.

**Comment by f00b4r0** [comment]
That doesn't work. Several boards share the same identifier and yet have different hardware capabilities. The correct "fix" is to not change the current value of RB_ID_BOARD_NAME (which breaks _all_ boards), and "simply" backport the LZOR extracting code. I'm still curious where that magic array `lzor_prefix` comes from. Maybe @robimarko can comment?

File: target/linux/ar71xx/files/arch/mips/ath79/routerboot.c (hunk 3 above)
Comment (f00b4r0) [reply target/linux/ar71xx/files/arch/mips/ath79/routerboot.c]: Just don't change RB_ID_BOARD_NAME, and remove that change entirely.
Besides, there really is no need to use memcmp to test for a null string 😛

**Comment by NewUse** [comment]
Thanks!, done.

**Comment by f00b4r0** [comment]
Thanks. This PR still imports a lot of code that is wrong in logic and suboptimal for a kernelspace implementation (coming straight from rbextract.c). The align change breaks the original intent (I suspect this is the reason why there are decoding errors in rbextract ("input not consumed")) and is implemented in an ugly way.

I am working on a clean implementation, which I suspect can be partially or totally backported to fix 19.07 if that's desired. I would welcome some input on the origins of the lzor_prefix blob which I am uncomfortable taking for granted.

HTH

**Comment by xback** [comment]
The only reason to keep this PR alive is to support the r2 revisions 19.07 so people dont run into boards lacking wifi for not so obvious reasons.

What i like about it is that is does minimal changes to the current state in 19.07 while adding r2 support.

Therefore I would be in favor to focus on implementing the really clean approach for Master only, while using this approach for 19.07.

I'll put this one in my staging tomorrow and test it on a huge nr of boards.

File: target/linux/ar71xx/files/arch/mips/ath79/routerboot.c [hunk 4]
```diff
@@ -98,20 +279,15 @@ routerboot_find_tag(u8 *buf, unsigned int buflen, u16 tag_id,
...
+		u32 id_and_len = get_u32(buf);
+		buf += 4;
+		buflen -= 4;
+		id = id_and_len & 0xFFFF;
+		len = id_and_len >> 16;
 
-		id = get_u16(buf);
-		buf += 2;
-		buflen -= 2;
+		if (align)
```
Comment (f00b4r0) [inline target/linux/ar71xx/files/arch/mips/ath79/routerboot.c]: This is wrong. Moving this block here will break lzo decoding as the length passed to the subroutine will be incorrect.

File: target/linux/ar71xx/files/arch/mips/ath79/routerboot.c [hunk 5]
```diff
@@ -120,17 +296,12 @@ routerboot_find_tag(u8 *buf, unsigned int buflen, u16 tag_id,
 			break;
 
 		if (id == tag_id) {
-			if (tag_len)
-				*tag_len = len;
-			if (tag_data)
-				*tag_data = buf;
+			*tag_len = len;
+			*tag_data = buf;
```
Comment (f00b4r0) [inline target/linux/ar71xx/files/arch/mips/ath79/routerboot.c]: This is wrong, the checks are needed. Expect NULL dereferences otherwise

File: target/linux/ar71xx/files/arch/mips/ath79/routerboot.c [hunk 6]
```diff
@@ -98,20 +279,15 @@ routerboot_find_tag(u8 *buf, unsigned int buflen, u16 tag_id,
...
+		buf += 4;
+		buflen -= 4;
+		id = id_and_len & 0xFFFF;
+		len = id_and_len >> 16;
 
-		id = get_u16(buf);
-		buf += 2;
-		buflen -= 2;
+		if (align)
+			len += (4 - len % 4) % 4;
```
Comment (f00b4r0) [inline target/linux/ar71xx/files/arch/mips/ath79/routerboot.c]: This is wrong. Use ALIGN() or the previous (less ugly) way of computing alignment.

File: target/linux/ar71xx/files/arch/mips/ath79/routerboot.c [hunk 7]
```diff
@@ -183,62 +354,143 @@ __rb_get_wlan_data(u16 id)
...
 
-	return buf;
+	}
+	return buf_rle_out;
 
 err_free:
-	kfree(buf);
+	kfree(buf_rle_out);
+	kfree(buf_lzo_out);
+	kfree(buf_lzo_in);
```
Comment (f00b4r0) [inline target/linux/ar71xx/files/arch/mips/ath79/routerboot.c]: All the intermediary buffers are never freed in normal operation

File: target/linux/ar71xx/files/arch/mips/ath79/routerboot.c [hunk 8]
```diff
@@ -183,62 +354,143 @@ __rb_get_wlan_data(u16 id)
...
+			goto err_free;
+		}
+
+		pr_err("Decompressing with LZO\n");
+		lzo_out_len = RB_ART_SIZE;
+		err = lzo1x_decompress_safe(buf_lzo_in, tag_len,
+					    buf_lzo_out, &lzo_out_len);
+		/* For some reason, I get this "input not consumed" error
+		 * even though the output is correct, so ignore it. */
+		if (err && err != LZO_E_INPUT_NOT_CONSUMED) {
```
Comment (f00b4r0) [inline target/linux/ar71xx/files/arch/mips/ath79/routerboot.c]: Of course this fails, tag_len is incorrect.

File: target/linux/ar71xx/files/arch/mips/ath79/routerboot.h [hunk 9]
```diff
@@ -8,6 +8,8 @@
  *  by the Free Software Foundation.
  */
 
+
+
```
Comment (f00b4r0) [inline target/linux/ar71xx/files/arch/mips/ath79/routerboot.h]: whitespace only changes throughout the file.

File: target/linux/generic/files/include/linux/routerboot.h [hunk 10]
```diff
@@ -42,6 +44,9 @@
 #define RB_ID_MAC_ADDRESS_COUNT	14
 #define RB_ID_HW_OPTIONS	21
 #define RB_ID_WLAN_DATA		22
+#define RB_ID_BOARD_IDENTIFIER	23
```
Comment (f00b4r0) [inline target/linux/generic/files/include/linux/routerboot.h]: Add this if necessary, but don't move BOARD_NAME

**Review by f00b4r0 — CHANGES_REQUESTED** [review]
Lots of problems

**Comment by f00b4r0** [comment]
> The only reason to keep this PR alive is to support the r2 revisions 19.07 so people dont run into boards lacking wifi for not so obvious reasons.
> 
> What i like about it is that is does minimal changes to the current state in 19.07 while adding r2 support.

I guess we have differing definitions of "minimal changes" 😛 
The blob alone is huge and unexplained.

> Therefore I would be in favor to focus on implementing the really clean approach for Master only, while using this approach for 19.07.

As I pointed out in my review, the changes as they are need more work.

> I'll put this one in my staging tomorrow and test it on a huge nr of boards.

Please don't merge this, I beg you 😿

**Comment by xback** [comment]
With "minimal changes" I mean not throwing away current way of working for a brand new sysfs approach. especially in a stable branch.

Don't panic, I'm not merging it as-is.
I've fixed most of your comments locally already ;-)

**Comment by f00b4r0** [comment]
> With "minimal changes" I mean not throwing away current way of working for a brand new sysfs approach. especially in a stable branch.

That's one way of using my code, another is to backport the fixed routines only (I have rewritten find_tag and the decoding function :)

> Don't panic, I'm not merging it as-is.
> I've fixed most of your comments locally already ;-)

😃

**Comment by xback** [comment]
Feel free to send me subsequent patches to this one (the original author also deserves credit) showing your changes.

In the end I only care about having the best possible and safest solution. :)

**Comment by f00b4r0** [comment]
It's a complete rewrite so it'll probably live in its own patchset (except for a backport proposal if that's desired). Don't worry, I always give credit where credit is due (been hacking the kernel for a few decades :)

**Comment by f00b4r0** [comment]
Speaking of which, this patchset doesn't credit the authors of the source code that's been copy-pasted. According to the header in rbextract.c:

```
 *  Copyright (C) 2012 Gabor Juhos <juhosg@openwrt.org>
 *  Copyright (C) 2018 Chris Schimp <silverchris@gmail.com>
 *  Copyright (C) 2019 Robert Marko <robimarko@gmail.com>
```

HTH

**Comment by f00b4r0** [comment]
@xback:

```
root@OpenWrt:~# ls /sys/firmware/mikrotik/hard_config/
board_identifier    board_serial        flash_info          mac_base            mem_size            wlan_data
board_product_code  booter_version      hw_options          mac_count           product_name
```

😉

**Comment by ae6xe** [comment]
Until " working on a clean implementation" and committed into 19.07, does this PR currently solve the problem and doesn't break something else?

File: target/linux/ar71xx/files/arch/mips/ath79/routerboot.c (hunk 4 above)
Comment (NewUse) [reply target/linux/ar71xx/files/arch/mips/ath79/routerboot.c:288]: 2 @f00b4r0 I am really bad in programming, I couldn't understand why placing here aligment condition could break lzo, if we are detecting RB_MAGIC_LZO lable in case condition align condition is ignored (align==false) so the lenght should be not changed?

Comment (f00b4r0) [reply target/linux/ar71xx/files/arch/mips/ath79/routerboot.c:288]: MAGIC_ERD uses lzo decompression. The comment below is in the `if (magic == RB_MAGIC_ERD) {` block.
Please note that robiemarko commented on the overall quality of the original code (from rbextract.c) here:
https://github.com/openwrt/openwrt/pull/2850#issuecomment-610277863

**Comment by NewUse** [comment]
revert most changes, need to be tested with new revisions.
code is still awful.

File: target/linux/generic/files/include/linux/routerboot.h (hunk 2 above)
Comment (nobbe64) [reply target/linux/generic/files/include/linux/routerboot.h]: IMG_20200417_142853 (1)

I have 3 of this boards. Is there any working Image yet (incuding wifi) and if yes can somebody tell me where I can get it from. If possible without patching and compiling

**Comment by adrianschmutzler** [comment]
The patch series from f00b4r0 has been merged. Despite, this PR would need to be revised for obvious formal issues by the author anyway.

@NewUse @xback Please check whether this should be updated or closed.

**Comment by f00b4r0** [comment]
This has become unneeded assuming fixing 18.06 is not desirable. If it is, it is probably cleaner to start from the 19.07 backport I'm preparing anyway. HTH

**Comment by NewUse** [comment]
It is unneeded now, so I am closing this PR.
waiting for backporting  f00b4r0 PR to 18.06

**Comment by xback** [comment]
@NewUse 

I still want to drop at thanks to you for initially opening this.
It certainly accelerated things in getting the initial issue fixed. :+1:
